File: reference/python/scor_ref/expr/lexer.py
# SPDX-FileCopyrightText: 2026 The Tessera Project
# SPDX-License-Identifier: Apache-2.0

"""Tokeniser for the hub expression DSL.

The grammar is deliberately small. No loops, no assignment, no I/O, no
string concatenation. If a token is not listed here it is not in the
language, and the parser will say so with a position.

Numbers and identifiers are ASCII-only. This is not cosmetic: Unicode
digits pass str.isdigit() but Decimal() rejects some of them, which used
to turn a malformed expression into an uncaught InvalidOperation instead
of a syntax error. The Rust implementation enforces the same rule, and
the conformance vectors pin it down for both.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List

from ..values import EvalError

KEYWORDS = {"and", "or", "not", "true", "false", "null"}

TWO_CHAR_OPS = {"<=", ">=", "==", "!="}
ONE_CHAR_OPS = set("+-*/<>(),")


@dataclass(frozen=True)
class Token:
    kind: str  # number | ident | keyword | op | end
    text: str
    pos: int


def tokenize(source: str) -> List[Token]:
    tokens: List[Token] = []
    i = 0
    n = len(source)
    while i < n:
        ch = source[i]
        if ch.isspace():
            i += 1
            continue
        if ch in "0123456789" or (ch == "." and i + 1 < n and source[i + 1] in "0123456789"):
            start = i
            seen_dot = False
            while i < n and (source[i] in "0123456789" or (source[i] == "." and not seen_dot)):
                if source[i] == ".":
                    seen_dot = True
                i += 1
            tokens.append(Token("number", source[start:i], start))
            continue
        if ch.isascii() and ch.isalpha() or ch == "_":
            start = i
            while i < n and ((source[i].isascii() and source[i].isalnum()) or source[i] in "_."):
                i += 1
            text = source[start:i]
            if text.endswith("."):
                raise EvalError("syntax", f"identifier ends with '.' at position {start}")
            kind = "keyword" if text in KEYWORDS else "ident"
            tokens.append(Token(kind, text, start))
            continue
        if source[i : i + 2] in TWO_CHAR_OPS:
            tokens.append(Token("op", source[i : i + 2], i))
            i += 2
            continue
        if ch in ONE_CHAR_OPS:
            tokens.append(Token("op", ch, i))
            i += 1
            continue
        raise EvalError("syntax", f"unexpected character {ch!r} at position {i}")
    tokens.append(Token("end", "", n))
    return tokens

File: rust/crates/scor-expr/src/lexer.rs
//! SPDX-FileCopyrightText: 2026 The Tessera Project
//! SPDX-License-Identifier: Apache-2.0

//! Tokeniser for the hub expression DSL.
//!
//! The grammar is deliberately small. No loops, no assignment, no I/O, no
//! string concatenation. If a token is not listed here it is not in the
//! language, and the parser will say so with a position.
//!
//! Numbers and identifiers are ASCII-only. This is not cosmetic: the Python
//! reference used to accept Unicode digits via `str.isdigit()` and then
//! crash inside `Decimal()` on some of them. Both implementations now
//! enforce ASCII, and the conformance vectors pin the rule down for both.

use std::fmt;

use super::values::EvalError;

/// Keywords reserved by the grammar.
pub const KEYWORDS: [&str; 6] = ["and", "or", "not", "true", "false", "null"];

/// Two-character operators.
pub const TWO_CHAR_OPS: [&str; 4] = ["<=", ">=", "==", "!="];

/// One-character operators.
pub const ONE_CHAR_OPS: [char; 9] = ['+', '-', '*', '/', '<', '>', '(', ')', ','];

/// The kind of a [`Token`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    /// A decimal literal.
    Number,
    /// A field reference or function name.
    Ident,
    /// A reserved keyword (`and`, `or`, `not`, `true`, `false`, `null`).
    Keyword,
    /// An operator or punctuation.
    Op,
    /// End of input.
    End,
}

/// A lexical token with its source position.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    /// Token kind.
    pub kind: TokenKind,
    /// Raw source text (empty for `End`).
    pub text: String,
    /// Byte offset of the token start in the source.
    pub pos: usize,
}

impl Token {
    /// The shared end-of-input token shape.
    #[must_use]
    pub fn end(pos: usize) -> Self {
        Self {
            kind: TokenKind::End,
            text: String::new(),
            pos,
        }
    }
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.text)
    }
}

/// Tokenise `source` into a token list terminated by [`TokenKind::End`].
///
/// # Errors
/// [`EvalError::Syntax`] on any character outside the grammar, or an
/// identifier that ends with `.`.
#[allow(
    clippy::indexing_slicing,
    reason = "the scan loop guards every access with i<n, the only lookahead uses chars.get(i+1), and text_of receives ranges the loop produced, all within bounds"
)]
pub fn tokenize(source: &str) -> Result<Vec<Token>, EvalError> {
    let chars: Vec<char> = source.chars().collect();
    let n = chars.len();
    let mut tokens: Vec<Token> = Vec::new();
    let mut i = 0;

    fn text_of(chars: &[char], from: usize, to: usize) -> String {
        chars[from..to].iter().collect()
    }

    while i < n {
        let ch = chars[i];
        if ch.is_whitespace() {
            i += 1;
            continue;
        }
        let starts_number = ch.is_ascii_digit()
            || (ch == '.' && chars.get(i + 1).is_some_and(char::is_ascii_digit));
        if starts_number {
            let start = i;
            let mut seen_dot = false;
            while i < n && (chars[i].is_ascii_digit() || (chars[i] == '.' && !seen_dot)) {
                if chars[i] == '.' {
                    seen_dot = true;
                }
                i += 1;
            }
            tokens.push(Token {
                kind: TokenKind::Number,
                text: text_of(&chars, start, i),
                pos: start,
            });
            continue;
        }
        if ch.is_ascii_alphabetic() || ch == '_' {
            let start = i;
            while i < n && (chars[i].is_ascii_alphanumeric() || chars[i] == '_' || chars[i] == '.')
            {
                i += 1;
            }
            let text = text_of(&chars, start, i);
            if text.ends_with('.') {
                return Err(EvalError::Syntax(format!(
                    "identifier ends with '.' at position {start}"
                )));
            }
            let kind = if KEYWORDS.contains(&text.as_str()) {
                TokenKind::Keyword
            } else {
                TokenKind::Ident
            };
            tokens.push(Token {
                kind,
                text,
                pos: start,
            });
            continue;
        }
        if i + 1 < n {
            let two = text_of(&chars, i, i + 2);
            if TWO_CHAR_OPS.contains(&two.as_str()) {
                tokens.push(Token {
                    kind: TokenKind::Op,
                    text: two,
                    pos: i,
                });
                i += 2;
                continue;
            }
        }
        if ONE_CHAR_OPS.contains(&ch) {
            tokens.push(Token {
                kind: TokenKind::Op,
                text: ch.to_string(),
                pos: i,
            });
            i += 1;
            continue;
        }
        return Err(EvalError::Syntax(format!(
            "unexpected character {ch:?} at position {i}"
        )));
    }
    tokens.push(Token::end(n));
    Ok(tokens)
}
